Index: include/llvm/MC/MCAssembler.h
===================================================================
--- include/llvm/MC/MCAssembler.h	(revision 152063)
+++ include/llvm/MC/MCAssembler.h	(working copy)
@@ -21,6 +21,10 @@
 #include "llvm/Support/DataTypes.h"
 #include <vector> // FIXME: Shouldn't be needed.
 
+namespace mcld {
+class Layout;
+}
+
 namespace llvm {
 class raw_ostream;
 class MCAsmLayout;
@@ -40,6 +44,7 @@
 
 class MCFragment : public ilist_node<MCFragment> {
   friend class MCAsmLayout;
+  friend class mcld::Layout;
 
   MCFragment(const MCFragment&);     // DO NOT IMPLEMENT
   void operator=(const MCFragment&); // DO NOT IMPLEMENT
@@ -53,7 +58,10 @@
     FT_Org,
     FT_Dwarf,
     FT_DwarfFrame,
-    FT_LEB
+    FT_LEB,
+    FT_Region,
+    FT_Reloc,
+    FT_Target
   };
 
 private:
@@ -679,7 +687,7 @@
 
   MCCodeEmitter &Emitter;
 
-  MCObjectWriter &Writer;
+  MCObjectWriter *Writer;
 
   raw_ostream &OS;
 
@@ -807,8 +815,10 @@
 
   MCCodeEmitter &getEmitter() const { return Emitter; }
 
-  MCObjectWriter &getWriter() const { return Writer; }
+  MCObjectWriter &getWriter() const { return *Writer; }
 
+  void setWriter(MCObjectWriter &ObjectWriter);
+
   /// Finish - Do final processing and write the object to the output stream.
   /// \arg Writer is used for custom object writer (as the MCJIT does),
   /// if not specified it is automatically created from backend.
Index: include/llvm/MC/MCSection.h
===================================================================
--- include/llvm/MC/MCSection.h	(revision 152063)
+++ include/llvm/MC/MCSection.h	(working copy)
@@ -31,7 +31,8 @@
     enum SectionVariant {
       SV_COFF = 0,
       SV_ELF,
-      SV_MachO
+      SV_MachO,
+      SV_LDContext
     };
 
   private:
Index: include/llvm/Support/ELF.h
===================================================================
--- include/llvm/Support/ELF.h	(revision 152063)
+++ include/llvm/Support/ELF.h	(working copy)
@@ -736,6 +736,13 @@
   SHT_LOOS          = 0x60000000, // Lowest operating system-specific type.
   SHT_HIOS          = 0x6fffffff, // Highest operating system-specific type.
   SHT_LOPROC        = 0x70000000, // Lowest processor architecture-specific type.
+
+  SHT_GNU_ATTRIBUTES = 0x6ffffff5, // Object attributes.
+  SHT_GNU_HASH       = 0x6ffffff6, // GNU style dynamic hash table.
+  SHT_GNU_verdef     = 0x6ffffffd, // Versions defined by file.
+  SHT_GNU_verneed    = 0x6ffffffe, // Versions needed by file.
+  SHT_GNU_versym     = 0x6fffffff, // Symbol versions.
+
   // Fixme: All this is duplicated in MCSectionELF. Why??
   // Exception Index table
   SHT_ARM_EXIDX           = 0x70000001U,
Index: lib/MC/MCAssembler.cpp
===================================================================
--- lib/MC/MCAssembler.cpp	(revision 152063)
+++ lib/MC/MCAssembler.cpp	(working copy)
@@ -157,7 +157,8 @@
 }
 
 MCFragment::MCFragment(FragmentType _Kind, MCSectionData *_Parent)
-  : Kind(_Kind), Parent(_Parent), Atom(0), Offset(~UINT64_C(0))
+  : Kind(_Kind), Parent(_Parent), Atom(0), Offset(~UINT64_C(0)),
+    LayoutOrder(~(0U))
 {
   if (Parent)
     Parent->getFragmentList().push_back(this);
@@ -197,7 +198,7 @@
 MCAssembler::MCAssembler(MCContext &Context_, MCAsmBackend &Backend_,
                          MCCodeEmitter &Emitter_, MCObjectWriter &Writer_,
                          raw_ostream &OS_)
-  : Context(Context_), Backend(Backend_), Emitter(Emitter_), Writer(Writer_),
+  : Context(Context_), Backend(Backend_), Emitter(Emitter_), Writer(&Writer_),
     OS(OS_), RelaxAll(false), NoExecStack(false), SubsectionsViaSymbols(false)
 {
 }
@@ -205,6 +206,11 @@
 MCAssembler::~MCAssembler() {
 }
 
+void MCAssembler::setWriter(MCObjectWriter &ObjectWriter) {
+  delete Writer;
+  Writer = &ObjectWriter;
+}
+
 bool MCAssembler::isSymbolLinkerVisible(const MCSymbol &Symbol) const {
   // Non-temporary labels should always be visible to the linker.
   if (!Symbol.isTemporary())
