Index: MCAssembler.h
===================================================================
--- MCAssembler.h	(revision 134957)
+++ MCAssembler.h	(working copy)
@@ -664,7 +664,7 @@
 
   MCCodeEmitter &Emitter;
 
-  MCObjectWriter &Writer;
+  MCObjectWriter *m_pWriter;
 
   raw_ostream &OS;
 
@@ -791,8 +791,13 @@
 
   MCCodeEmitter &getEmitter() const { return Emitter; }
 
-  MCObjectWriter &getWriter() const { return Writer; }
+  MCObjectWriter &getWriter() const { return *m_pWriter; }
 
+  void setWriter(MCObjectWriter &pObjectWriter) {
+    delete m_pWriter;
+    m_pWriter = &pObjectWriter;
+  }
+
   /// Finish - Do final processing and write the object to the output stream.
   /// \arg Writer is used for custom object writer (as the MCJIT does),
   /// if not specified it is automatically created from backend.
