Index: TargetMachine.h
===================================================================
--- TargetMachine.h	(revision 134892)
+++ TargetMachine.h	(working copy)
@@ -311,7 +311,7 @@
   LLVMTargetMachine(const Target &T, StringRef TargetTriple,
                     StringRef CPU, StringRef FS);
 
-private:
+public:
   /// addCommonCodeGenPasses - Add standard LLVM codegen passes used for
   /// both emitting to assembly files or machine code output.
   ///
